src/ocean/net/ssl/SslClientConnection.d

```diff
@@ -0,0 +1,622 @@
+
+
+    /***************************************************************************
+
+        Create an SSL connection. Blocks the calling task until the handshake
+        is complete.
+
+        Params:
+            host_name = the name of the host
+            host_port = the port to use (eg "443" for an HTTPS connection)
+
+        Throws:
+            IOException if the connection failed
+
+    ***************************************************************************/
+
+    public void connect ( cstring host_name, cstring host_port )
+    {
+        // Resolve the host
+
+        // We should be able to use AddrInfo.getIp but cannot because it doesn't
+        // support AF_UNSPEC.
+
+        auto addr_err = this.addr_info.get(host_name, host_port, AF_UNSPEC,
+            SOCK_STREAM, 0);
+
+        if ( addr_err != 0 )
+        {
+            this.ssl_error(host_name, "connect", gai_strerror(addr_err));
+        }
+
+        this.socket.socket(addr_info.info().ai_family,
+            addr_info.info().ai_socktype | SocketFlags.SOCK_NONBLOCK,
+            addr_info.info().ai_protocol);
+
+        if ( !socket.connect(addr_info.info().ai_addr) )
+        {
+            this.error(host_name, "socket.connect", "Failed to connect");
+        }
+
+        this.sslHandle = SSL_new(globalSslContext);
+
+        if ( !this.sslHandle )
+        {
+            this.error(host_name, "connect", "sslNew failed");
+        }
+
+        if (!SSL_set_fd(this.sslHandle, this.socket.fileHandle))
+        {
+            this.error(host_name, "connect", "sslSetFd failed");
+        }
+
+        // Set it into "connect" state, ie it is a client, not a server.
+        // (For a server, SSL_set_accept_state would be called instead).
+
+        SSL_set_connect_state(this.sslHandle);
+
+        ERR_clear_error();
+
+        int r;
+
+        while ( (r = SSL_do_handshake(this.sslHandle)) != 1 )
+        {
+            int err = SSL_get_error(this.sslHandle, r);
+
+            if ( err ==  SSL_ERROR_WANT_WRITE )
+            {
+                this.select_client.ioWait(Event.EPOLLOUT);
+            }
+            else if (err == SSL_ERROR_WANT_READ)
+            {
+                this.select_client.ioWait(Event.EPOLLIN);
+            }
+            else
+            {
+                this.error(host_name, "connect", "Handshake failed");
+            }
+        }
+    }
+
+
+    /***************************************************************************
+
+        Write a string to the SSL connection
+
+        The calling task is blocked until the write has completed.
+
+        Params:
+            request = the string to send
+
+        Throws:
+            IOException if the write failed
+
+    ***************************************************************************/
+
+    public void write ( cstring request )
+    {
+        int w;
+
+        while ( (w = SSL_write(this.sslHandle, cast(void *)request.ptr,
+            cast(int)request.length)) <= 0 )
```

Comment:
Casting `const` away from `request.ptr` deserves a note.

Reply:
No need to cast at all, just declare `SSL_write` as taking const argument.